services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java

```diff
@@ -304,6 +306,48 @@ protected SimpleHttp getRefreshTokenRequest(KeycloakSession session, String refr
         return authenticateTokenRequest(refreshTokenRequest);
     }
 
+    @Override
+    public Response retrieveToken(KeycloakSession session, FederatedIdentityModel identity) {
```

Comment:
The `Automatic refresh of external IDP token` is actually already supported for the internal-external token exchange. It is implemented in the method `OIDCIdentityProvider.exchangeStoredToken` . Could we possibly make sure that both places are using same logic (and maybe re-using same code if possible) ?

For example `exchangeStoredToken` takes into consideration `getConfig().getMinValidityToken()` while it seems that this one does not.

Reply:
I aligned the logic, and extracted two methods. 

I didn't use them with the existing code yet, I would like to keep it in a separate PR as it would lead to too many changes. And I would like to merge #31275 first which fixes another bug and adds more tests.